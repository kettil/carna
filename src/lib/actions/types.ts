import { Watermark, Watermarks } from 'istanbul-lib-report';
import { SpawnKillHandler } from '../utils/createSpawnKillHandler';

type BabelActionProps = {
  watch?: boolean;
  spawnKillHandler?: SpawnKillHandler;
};

type BabelNodeActionProps = {
  watch: boolean;
  script: string;
  watchPaths?: string[];
};

type CommitlintActionProps = {
  edit: string;
};

type CoverageWatermarkThreshold = Watermark | number | undefined;
type CoverageActionProps = {
  projects: string[];
  watermarks?: Partial<Record<keyof Watermarks, CoverageWatermarkThreshold>>;
};

type DepcheckActionProps = {
  path: string;
  ignorePackages?: string[];
};

type EslintActionProps = {
  write?: boolean;
  files?: string[];
};

type JestActionProps = {
  projects: string[];
  updateSnapshot?: boolean;
  runInBand?: boolean;
  verbose?: boolean;
  watch?: boolean;
};

type LicensecheckActionProps = {
  ignorePackages: string[];
  path?: string;
};

type PrettierActionProps = {
  write?: boolean;
  extension?: string;
  files?: string[];
};

type SemverActionProps = {
  path?: string;
};

type TscActionProps = {
  mode: 'type-check' | 'type-create';
};

export type {
  BabelActionProps,
  BabelNodeActionProps,
  CommitlintActionProps,
  CoverageActionProps,
  CoverageWatermarkThreshold,
  DepcheckActionProps,
  EslintActionProps,
  JestActionProps,
  LicensecheckActionProps,
  PrettierActionProps,
  SemverActionProps,
  TscActionProps,
};
