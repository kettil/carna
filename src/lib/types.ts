import { Logger } from './cli/logger';

export type NpmInstallMode = 'prod' | 'dev' | 'optional';

export type PropsGlobal = {
  readonly cwd: string;
  readonly tpl: string;
  readonly cfg: string;
  readonly log: Logger;
  readonly vvv: boolean;
};

export type Action<
  Props extends Record<string, unknown> | undefined = undefined,
  ActionReturn = void
> = Props extends undefined
  ? (argv: PropsGlobal) => Promise<ActionReturn>
  : (argv: PropsGlobal, props: Props) => Promise<ActionReturn>;
