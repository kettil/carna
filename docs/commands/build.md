# The `build` task

Build the application

## Options

| Options | Description       |
| ------- | ----------------- |
| --ci    | Run it in CI mode |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
