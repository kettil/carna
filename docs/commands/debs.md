# The `debs` task

Checks if there are orphaned packages

## Options

| Options | Description       |
| ------- | ----------------- |
| --ci    | Run it in CI mode |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
