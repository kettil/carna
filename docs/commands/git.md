# The `git` tasks

The tasks will be automatically called by husky during commit.

## The `git commit` task

Checks the files in staged with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git msg` task

Checks the git commit message for a consistent structure.

### Options

| Options | Description                     |
| ------- | ------------------------------- |
| --edit  | Path to the COMMIT_EDITMSG file |

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git push` task

Checks the files with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
