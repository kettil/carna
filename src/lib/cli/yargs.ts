import { CommandModule, Arguments, Argv } from 'yargs';
import ExecutableError from '../errors/executableError';
import { PropsGlobal } from '../types';

export type CommandModuleCommand = string;

export type CommandModuleDescribe = NonNullable<CommandModule<PropsGlobal>['describe']>;

export type CommandModuleBuilder<Props = Record<string, unknown>> = (
  yargs: Argv<PropsGlobal>,
) => Argv<PropsGlobal & Props>;

export type CommandModuleHandler<Props = Record<string, unknown>> = (args: Arguments<PropsGlobal & Props>) => void;

export const builderDefault = <Props>(cmd: string, fn: CommandModuleBuilder<Props>): CommandModuleBuilder<Props> => (
  yargs,
) => fn(yargs).usage(`Usage: $0 ${cmd} [options]`).help().version(false)
  .showHelpOnFail(false);

export const errorHandler = (argv: PropsGlobal, error: unknown): void => {
  if (error instanceof ExecutableError) {
    argv.log.log(error.stdout);

    /* eslint-disable-next-line node/no-process-exit, unicorn/no-process-exit */
    process.exit(1);
  }

  throw error;
};
