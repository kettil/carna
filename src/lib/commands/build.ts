import { createBuilder, createHandler } from '../cli/yargs';
import { buildTask, BuildProps } from '../tasks/buildTask';

const command = 'build';
const desc = 'Run the build process';

const options = { group: `${command}-Options` } as const;
const boolOptions = { ...options, type: 'boolean', default: false } as const;

const handler = createHandler<BuildProps>(buildTask);
const builder = createBuilder<BuildProps>(command, (yargs) =>
  yargs.options({
    watch: {
      ...boolOptions,
      alias: 'w',
      describe: 'Watch files for changes and rebuild the changed files (without type files)',
    },
  }),
);

export { command, desc, builder, handler };
