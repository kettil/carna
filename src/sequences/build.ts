import npmPackageLoad from '../actions/npm/packageLoad';
import babel from '../actions/tools/babel';
import tsc from '../actions/tools/tsc';
import webpack from '../actions/tools/webpack';
import { spinnerAction } from '../lib/cli/spinner';
import {
  CommandModuleBuilder,
  CommandModuleDescribe,
  CommandModuleCommand,
  CommandModuleHandler,
  builderDefault,
  errorHandler,
  ciDefaultValue,
  commonHandler,
} from '../lib/cli/yargs';

export const command: CommandModuleCommand = 'build';
export const desc: CommandModuleDescribe = 'Run the code quality tools';

const options = { group: `${command}-Options` } as const;

type Props = {
  readonly ci: boolean;
};

export const builder: CommandModuleBuilder<Props> = builderDefault(command, (yargs) =>
  yargs.options({
    ci: { ...options, type: 'boolean', default: ciDefaultValue(), describe: 'Run it in CI mode' },
  }),
);

export const handler: CommandModuleHandler<Props> = async (argv) => {
  try {
    await commonHandler(argv, !argv.ci);

    const isPrivate = await npmPackageLoad(argv, { key: 'private' });

    if (isPrivate !== true) {
      await spinnerAction(tsc(argv, { mode: 'type-create' }), 'Typescript');
    }

    await spinnerAction(babel(argv), 'Babel');
    await spinnerAction(webpack(argv), 'Webpack');
  } catch (error) {
    errorHandler(argv, error);
  }
};
