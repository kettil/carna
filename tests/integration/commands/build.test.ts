import * as fs from 'fs';
import { Argv } from 'yargs';
import { handler, builder, command, desc } from '../../../src/lib/commands/build';
import { PropsGlobal } from '../../../src/lib/types';
import { getArgv } from '../../shared/setup/argv';
import { mockFilePackage, packageJsonWithWorkspaces } from '../../shared/setup/packageJson';
import { getReadFileFiles, getReadFileWithHooksFiles } from '../../shared/setup/readFileFiles';

describe('command build', () => {
  test('it should be complete the yargs command structure', () => {
    expect(typeof command).toBe('string');
    expect(typeof desc).toBe('string');
    expect(typeof builder).toBe('function');
    expect(typeof handler).toBe('function');
  });

  test('it should be create the expected builder', () => {
    expect.assertions(2);

    const yargs = {
      options: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('options');

        return yargs;
      }),
      usage: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('usage');

        return yargs;
      }),

      help: jest.fn(() => yargs),
      version: jest.fn(() => yargs),
      showHelpOnFail: jest.fn(() => yargs),
    } as unknown as Argv<PropsGlobal>;

    builder(yargs);
  });

  test('it should work', async () => {
    const result = await handler(getArgv({}));

    expect(result).toBeUndefined();
  });

  test('it should work with hooks', async () => {
    (fs as any).setMockReadFileFiles(getReadFileWithHooksFiles());

    const result = await handler(getArgv({}));

    expect(result).toBeUndefined();
  });

  test.each([[false], [true]])('it should work with workspaces [watch: %p]', async (watch) => {
    (fs as any).setMockReadFileFiles(
      getReadFileFiles({
        [mockFilePackage]: JSON.stringify(packageJsonWithWorkspaces),
      }),
    );

    const result = await handler(getArgv({ watch }));

    expect(result).toBeUndefined();
  });
});
