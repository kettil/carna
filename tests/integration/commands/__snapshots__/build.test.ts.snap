// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command build it should be create the expected builder: options 1`] = `
Array [
  Object {
    "watch": Object {
      "alias": "w",
      "default": false,
      "describe": "Watch files for changes and rebuild the changed files (without type files)",
      "group": "build-Options",
      "type": "boolean",
    },
  },
]
`;

exports[`command build it should be create the expected builder: usage 1`] = `
Array [
  "Usage: $0 build [options]",
]
`;

exports[`command build it should work with hooks: child-process 1`] = `
Object {
  "args": Array [
    "run",
    "precarna:build",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command build it should work with hooks: child-process 2`] = `
Object {
  "args": Array [
    "--outDir",
    "/path/to/project/build",
    "--project",
    "/path/to/project/tsconfig.build.json",
    "--noEmit",
    "false",
    "--emitDeclarationOnly",
  ],
  "command": "/path/to/project/node_modules/.bin/tsc",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command build it should work with hooks: child-process 3`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command build it should work with hooks: child-process 4`] = `
Object {
  "args": Array [
    "run",
    "postcarna:build",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command build it should work with workspaces [watch: false]: child-process 1`] = `
Object {
  "args": Array [
    "--outDir",
    "/path/to/project/packages/a/build",
    "--project",
    "/path/to/project/packages/a/tsconfig.json",
    "--noEmit",
    "false",
    "--emitDeclarationOnly",
  ],
  "command": "/path/to/project/node_modules/.bin/tsc",
  "cwd": "/path/to/project/packages/a",
  "env": Object {},
}
`;

exports[`command build it should work with workspaces [watch: false]: child-process 2`] = `
Object {
  "args": Array [
    "--outDir",
    "/path/to/project/packages/b/build",
    "--project",
    "/path/to/project/packages/b/tsconfig.json",
    "--noEmit",
    "false",
    "--emitDeclarationOnly",
  ],
  "command": "/path/to/project/node_modules/.bin/tsc",
  "cwd": "/path/to/project/packages/b",
  "env": Object {},
}
`;

exports[`command build it should work with workspaces [watch: false]: child-process 3`] = `
Object {
  "args": Array [
    "--outDir",
    "/path/to/project/packages/c/build",
    "--project",
    "/path/to/project/packages/c/tsconfig.json",
    "--noEmit",
    "false",
    "--emitDeclarationOnly",
  ],
  "command": "/path/to/project/node_modules/.bin/tsc",
  "cwd": "/path/to/project/packages/c",
  "env": Object {},
}
`;

exports[`command build it should work with workspaces [watch: false]: child-process 4`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project/packages/a",
  "env": Object {},
}
`;

exports[`command build it should work with workspaces [watch: false]: child-process 5`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project/packages/b",
  "env": Object {},
}
`;

exports[`command build it should work with workspaces [watch: false]: child-process 6`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project/packages/c",
  "env": Object {},
}
`;

exports[`command build it should work with workspaces [watch: false]: fs 1`] = `
Object {
  "dest": "/path/to/project/packages/a/.npmignore",
  "src": "/path/to/project/.npmignore",
}
`;

exports[`command build it should work with workspaces [watch: false]: fs 2`] = `
Object {
  "dest": "/path/to/project/packages/b/.npmignore",
  "src": "/path/to/project/.npmignore",
}
`;

exports[`command build it should work with workspaces [watch: false]: fs 3`] = `
Object {
  "dest": "/path/to/project/packages/c/.npmignore",
  "src": "/path/to/project/.npmignore",
}
`;

exports[`command build it should work with workspaces [watch: false]: glob 1`] = `
Object {
  "options": Object {
    "absolute": true,
    "cwd": "/path/to/project",
  },
  "pattern": "packages/*",
}
`;

exports[`command build it should work with workspaces [watch: true]: child-process 1`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "--watch",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project/packages/a",
  "env": Object {},
}
`;

exports[`command build it should work with workspaces [watch: true]: child-process 2`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "--watch",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project/packages/b",
  "env": Object {},
}
`;

exports[`command build it should work with workspaces [watch: true]: child-process 3`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "--watch",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project/packages/c",
  "env": Object {},
}
`;

exports[`command build it should work with workspaces [watch: true]: glob 1`] = `
Object {
  "options": Object {
    "absolute": true,
    "cwd": "/path/to/project",
  },
  "pattern": "packages/*",
}
`;

exports[`command build it should work with workspaces [watch: true]: process-stdin 1`] = `
Object {
  "mode": true,
}
`;

exports[`command build it should work with workspaces [watch: true]: process-stdin 2`] = `
Object {
  "encoding": "utf8",
}
`;

exports[`command build it should work with workspaces [watch: true]: process-stdin 3`] = `
Object {
  "callback": [Function],
  "event": "data",
}
`;

exports[`command build it should work: child-process 1`] = `
Object {
  "args": Array [
    "--outDir",
    "/path/to/project/build",
    "--project",
    "/path/to/project/tsconfig.build.json",
    "--noEmit",
    "false",
    "--emitDeclarationOnly",
  ],
  "command": "/path/to/project/node_modules/.bin/tsc",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command build it should work: child-process 2`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;
